src/pds/registrysweepers/ancestry.py

```diff
@@ -0,0 +1,259 @@
+    docs = query_f(host, query, _source)  # type: ignore
+
+    return [AncestryRecord(lidvid=PdsLidVid.from_string(doc["_source"]["lidvid"])) for doc in docs]
+
+
+def get_collection_ancestry_records(
+    host: HOST, registry_db_mock: Optional[Callable[[str], Iterable[Dict]]] = None
+) -> Iterable[AncestryRecord]:
+    # Query the registry for all bundles and the collections each references.
+    bundles_query = product_class_query_factory(ProductClass.BUNDLE)
+    bundles_source = {"includes": ["lidvid", "ref_lid_collection"]}
+    bundles_query_f = query_registry_db_or_mock(registry_db_mock, "get_collection_ancestry_records_bundles")
+    bundles_docs = bundles_query_f(host, bundles_query, bundles_source)  # type: ignore
+
+    # Query the registry for all collection identifiers
+    collections_query = product_class_query_factory(ProductClass.COLLECTION)
+    collections_source = {"includes": ["lidvid"]}
+    # TODO: switch to this line to support alternate_ids
+    # collection_identifiers_source = {"includes": ["alternate_ids"]}  # alternate_ids includes lid, lidvid, and any aliases
+    collections_query_f = query_registry_db_or_mock(registry_db_mock, "get_collection_ancestry_records_collections")
+    collection_identifiers_docs = collections_query_f(host, collections_query, collections_source)  # type: ignore
+
+    # TODO: change to for loop to support alternate_ids (generate multiple keys per element and filter out LIDs)
+    #  also necessary to populate a multidirectional lookup for LIDs with aliases which can be used during LID-based
+    #  reference assignment
+    # Instantiate the AncestryRecords, keyed by collection LIDVID for fast access
+    ancestry_by_collection_lidvid: Dict[PdsLidVid, AncestryRecord] = {
+        PdsLidVid.from_string(doc["_source"]["lidvid"]): AncestryRecord(
+            lidvid=PdsLidVid.from_string(doc["_source"]["lidvid"])
+        )
+        for doc in collection_identifiers_docs
+    }
+
+    # Create a dict of pointer-sets to the newly-instantiated records, binned/keyed by LID for fast access when a bundle
+    #  only refers to a LID rather than a specific LIDVID
+    ancestry_by_collection_lid: Dict[PdsLid, Set[AncestryRecord]] = {}
+    for record in ancestry_by_collection_lidvid.values():
+        if record.lidvid.lid not in ancestry_by_collection_lid:
+            ancestry_by_collection_lid[record.lidvid.lid] = set()
+        ancestry_by_collection_lid[record.lidvid.lid].add(record)
+
+    # For each bundle, add it to the bundle-ancestry of every collection it references
+    for doc in bundles_docs:
+        bundle_lidvid = PdsLidVid.from_string(doc["_source"]["lidvid"])
+        referenced_collection_identifiers = [
+            PdsProductIdentifierFactory.from_string(id) for id in doc["_source"]["ref_lid_collection"]
+        ]
+
+        for identifier in referenced_collection_identifiers:
+            if identifier.__class__ is PdsLidVid:
```

Comment:
Let me have a think about this - the suggested solution isn't desirable because `PdsProductIdentifier`s (of any stripe) exist in isolation from any particular use-case, and I'd rather not couple them.  The second I see another python repo that could use these classes I plan to extract them into a pip-installable package.  You could have an abstract interface class `Ancestry` requiring implementation of that method and add subclasses of PdsLid/PdsLidVid which also inherit it, but... that's ten times as nasty and hard to justify beyond cargo-cult adherence to OOP (which the sweepers aren't, even - they're procedural, and just leverage data-objects where useful).

This is kind of a special case, because the OOP principle that classes should be open to extension isn't actually true for `PdsProductIdentifier`, which will only ever have direct-subclasses `PdsLid` and `PdsLidVid`, _ever_ (or at least, for the lifetime of PDS4).

That being said, there's nothing to say that _they_ couldn't be subclassed for some purpose, so the use of `identifier.__class__ is PdsLidVid` instead of `isinstance(identifier, PdsLidVid)` is a straight-up mistake I'll need to fix.